cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1401,13 +1412,27 @@ private AppFabricServiceStatus stop(Id.Program identifier, ProgramType type) {
     }
   }
 
-  private void getRuns(HttpResponder responder, Id.Program programId, String status,
-                       long start, long end, int limit) {
+  private void getRuns(HttpResponder responder, Id.Program programId, String status, long start, long end, int limit) {
     try {
       try {
-        ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
+        final ProgramType programType = programId.getType();
+        final ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
           ProgramRunStatus.valueOf(status.toUpperCase());
-        responder.sendJson(HttpResponseStatus.OK, store.getRuns(programId, runStatus, start, end, limit));
+
+        // Need to validate whether the run records are reflecting the actual running state of the program.
+        List<RunRecord> runRecords = lifecycleService.validateRunRecords(
+          store.getRuns(programId, runStatus, start, end, limit), programId, programType);
```

Comment:
What I meant was that this endpoint can be used to show all runs of a particular run status. It is perfectly valid to call this to list all failed runs. With your change, the runs that have failed but have not been updated as failed in the run record will not be listed. This will lead to perceived inconsistencies. Some runs will show as failed when listing ALL, but would not show up when listing FAILED. Or, if you list FAILED ones first, RUNNING ones next, you will not see some runs. But if you first list RUNNING runs, then FAILED ones, these runs will show as failed.